### PR TITLE
Make `__precompile__(false)` non-fatal for cachecompile / requires

### PR DESCRIPTION
**Pull request by vtjnash**
replaces #28386

### CODE REVIEW

File: stdlib/Pkg/src/API.jl [hunk 1]
```diff
@@ -445,52 +445,33 @@ precompile() = precompile(Context())
...
+            break
         end
-        if !found_matching_precompile
-            # Don't bother attempting to precompile packages that appear to contain `__precompile__(false)`
-            source = read(sourcepath, String)
-            if !occursin(r"__precompile__\(false\)", source)
-                push!(needs_to_be_precompiled, pkg.name)
-            end
+        if stale
+            printpkgstyle(ctx, :Precompiling, pkg)
```
Comment (ararslan) [inline stdlib/Pkg/src/API.jl]: `pkg` is a `PkgId` but this function expects the third argument to be a string, which is why the tests are failing.

**Comment by StefanKarpinski** [comment]
This is the last fix blocking the next RC.

**Comment by Keno** [comment]
Fixed and rebased.

File: test/precompile.jl [hunk 2]
```diff
@@ -279,18 +280,15 @@ try
...
-    @test_warn "ERROR: LoadError: Declaring __precompile__(false) is not allowed in files that are being precompiled.\nStacktrace:\n [1] __precompile__" try
-        Base.compilecache(Base.PkgId("Baz")) # from __precompile__(false)
-        error("__precompile__ disabled test failed")
-    catch exc
-        isa(exc, ErrorException) || rethrow(exc)
-        occursin("__precompile__(false)", exc.msg) && rethrow(exc)
-    end
+    @test Base.compilecache(Base.PkgId("Baz")) == Base.PrecompilableError() # due to __precompile__(false)
+    @eval using Baz
+    @test Base.invokelatest(Baz.baz) == 1
```
Comment (ngam) [inline test/precompile.jl:291]: Something in 1.7.0 is breaking this on some machines.

Comment (ngam) [reply test/precompile.jl:291]: @ararslan
@Keno 
@StefanKarpinski 

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=432259&view=logs&j=656edd35-690f-5c53-9ba3-09c10d0bea97&t=e5c8ab1d-8ff9-5cae-b332-e15ae582ed2d&l=4061

**Pull request merged**